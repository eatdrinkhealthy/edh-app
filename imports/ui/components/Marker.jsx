// @flow
import React, {
  PureComponent,
} from "react";
import classNames from "classnames";

type IMarkerOrigin =
  | "center"
  | "topLeft"
  | "bottomCenter";

type IMarkerProps = {
  label?: string,
  origin?: IMarkerOrigin,
};

type IMarkerDefaultProps = {
  origin: IMarkerOrigin,
};

class Marker extends PureComponent {
  static defaultProps: IMarkerDefaultProps = {
    origin: "bottomCenter",
  };

  props: IMarkerProps;

  render() {  // eslint-disable-line flowtype/require-return-type
    const markerClass = classNames("markerStyle", {
      markerOriginCenter: this.props.origin === "center",
      markerOriginTopLeft: this.props.origin === "topLeft",
      markerOriginBottomCenter: this.props.origin === "bottomCenter",
    });

    return (
      <div className={markerClass}>
        {this.props.label}
        <img src="images/pinMarker3x.png" alt="pin marker" className="markerImage" />
      </div>
    );
  }
}

export default Marker;
