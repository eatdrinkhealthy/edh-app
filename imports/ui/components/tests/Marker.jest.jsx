// @flow
/* eslint-env jest */
/* eslint-disable func-names, prefer-arrow-callback, no-unused-expressions */
/* eslint-disable import/no-extraneous-dependencies */
import React from "react";
import { shallow } from "enzyme";
import renderer from "react-test-renderer";
import Marker, { calcHintPosition } from "../Marker";
import sampleVenues from "../../../data/state/stores/tests/sampleVenueData";

describe("<Marker />", function () {
  // Per a React blog post, when using renderer, must mock out refs
  // https://facebook.github.io/react/blog/2016/11/16/react-v15.4.0.html#mocking-refs-for-snapshot-testing
  // eslint-disable-next-line flowtype/no-weak-types
  function createNodeMock(element: HTMLElement): ?Object {
    if (element.type === "div") {
      return {
        getBoundingClientRect() {},
      };
    }
    return null;
  }

  const options = { createNodeMock };

  const testVenue = { ...sampleVenues[0] };  // create a copy of a sample venue object

  it("matches render snapshot, unselected", function () {
    const tree = renderer.create(<Marker
      venue={testVenue}
      setSelectedVenueHandler={() => {}}
    />, options).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("matches render snapshot, selected", function () {
    const tree = renderer.create(<Marker
      venue={testVenue}
      selected
      setSelectedVenueHandler={() => {}}
    />, options).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should have default style and origin classes", function () {
    const wrapper = shallow(<Marker
      venue={testVenue}
      setSelectedVenueHandler={() => {}}
    />);
    expect(wrapper.hasClass("markerContainer")).toBe(true);
    expect(wrapper.hasClass("markerOriginBottomCenter")).toBe(true);
  });

  it("should call setSelectedVenueHandler when clicked", function () {
    const props = {
      venue: testVenue,
      setSelectedVenueHandler: jest.fn(),
    };

    const wrapper = shallow(<Marker {...props} />);
    wrapper.find("div.markerContainer").simulate("click");
    expect(props.setSelectedVenueHandler.mock.calls.length).toBe(1);
    expect(props.setSelectedVenueHandler.mock.calls[0][0]).toBe("1");
  });

  it("should have class 'hint--bottom' if no viewBoundaryRect prop is passed", function () {
    const wrapper = shallow(<Marker
      venue={testVenue}
      setSelectedVenueHandler={() => {}}
    />);

    expect(wrapper.hasClass("hint--bottom")).toBe(true);
  });

  describe("calcHintPosition", function () {
    it("should return 'hint--bottom' if viewBoundaryRect is null", function () {
      expect(calcHintPosition(null)).toBe("hint--bottom");
    });
  });
});
