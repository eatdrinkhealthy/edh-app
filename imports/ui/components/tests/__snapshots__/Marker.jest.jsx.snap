// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Marker /> matches render snapshot, no text 1`] = `
<div
  className="markerStyle markerOriginBottomCenter"
>
  <img
    alt="pin marker"
    className="markerImage"
    src="images/pinMarker3x.png"
  />
</div>
`;

exports[`<Marker /> matches render snapshot, with text 1`] = `
<div
  className="markerStyle markerOriginBottomCenter"
>
  Some Place Cool
  <img
    alt="pin marker"
    className="markerImage"
    src="images/pinMarker3x.png"
  />
</div>
`;
